app/routes.py

```diff
@@ -1 +1,298 @@
+    
+    tasks.title = request_data["title"]
+    tasks.description = request_data["description"]
+    tasks.completed_at = request_data.get("completed_at", None)
+
+    db.session.commit()
+
+    return jsonify({"task":{
+            "id": tasks.task_id,
+            "title": tasks.title,
+            "description": tasks.description,
+            "is_complete": False
+            }
+            }), 200
+
+
+# delete task
+@task_bp.route("/<task_id>", methods=["DELETE"])
+def delete_task(task_id):
+    tasks = validate_task(Task,task_id)
+
+    db.session.delete(tasks)
+    db.session.commit()
+
+    return {"details": f'Task {task_id} "Go on my daily walk 🏞" successfully deleted'}
+
+
+# helper function
+def validate_task(model,task_id):
+    try:
+        task_id = int(task_id)
+    except ValueError:
+        return abort(make_response({"msg": "id is invalide input"}, 400))
+    
+    # return Task.query.get_or_404(task_id_num)
+ 
+    task = model.query.get(task_id)
+    if task is None:
+        abort(make_response({"msg": "Task not found"}, 404))
+    return task
+
+
+# wave 2 
+# get task by asc and desc order
+@task_bp.route("", methods =["GET"])
+def sort_title_asc_and_desc():
+    sort_by = request.args.get("sort")
+
+    if sort_by == "asc":
+        tasks = Task.query.order_by(Task.title.asc()).all()
+    elif sort_by == "desc":
+        tasks = Task.query.order_by(Task.title.desc()).all()
+    else:
+        tasks = Task.query.all() # this line could save one function for get_all_tasks
+    
+
+    response =[task.to_dict() for task in tasks]
+
+    return jsonify(response), 200
+
+
+# wave 3 
+# mark complete
+@task_bp.route("/<task_id>/mark_complete", methods = ["PATCH"])
+def mark_complete(task_id):
+
+    task = Task.query.get(task_id)
+    if not task:
+        return jsonify({"msg": "task not found"}), 404
+    
+    if not task.completed_at:
+        task.completed_at = datetime.utcnow()
+        task.is_complete = True 
+
+        db.session.commit()
+
+        url = "https://slack.com/api/chat.postMessage"
+        token = os.environ.get("SLACKBOT_TOKEN")
+        data ={"channel": "task-notifications",
+               "text": f"Someone just completed a task {task.title}",
+               "token": token
+               }
+        response = requests.post(url, data = data)
+
+
+        # this is the part with slack module:
+        # message = f"Someone just completed a task {task.title}"
+        # client = WebClient(token=os.environ.get("SLACKBOT_TOKEN"))
+        # response = client.chat_postMessage(channel="task-notifications", text=message)
+
+         
+    return jsonify({"task":task.to_dict()}), 200
+
+
+# mark_imcomplete
+@task_bp.route("/<task_id>/mark_incomplete", methods = ["PATCH"])
+def mark_imcomplete(task_id):
+    task = Task.query.get(task_id)
+    if not task:
+        return jsonify({"msg": "task not found"}), 404
+    
+    task.completed_at = None
+    task.is_complete = False
```

Comment:
This line is not necessary. The Task class does not have an is_complete attribute.